tower-http/src/compression/body.rs

```diff
@@ -14,124 +14,216 @@ use bytes::{Buf, Bytes};
 use futures_util::ready;
 use http::HeaderMap;
 use http_body::Body;
-use pin_project::pin_project;
+use pin_project_lite::pin_project;
 use std::{
     io,
+    marker::PhantomData,
     pin::Pin,
     task::{Context, Poll},
 };
 use tokio_util::io::StreamReader;
 
-/// Response body of [`Compression`].
-///
-/// [`Compression`]: super::Compression
-#[pin_project]
-pub struct CompressionBody<B>(#[pin] pub(crate) BodyInner<B>)
-where
-    B: Body;
+pin_project! {
+    /// Response body of [`Compression`].
+    ///
+    /// [`Compression`]: super::Compression
+    pub struct CompressionBody<B>
+    where
+        B: Body,
+    {
+        #[pin]
+        pub(crate) inner: BodyInner<B>,
+    }
+}
 
 impl<B> CompressionBody<B>
 where
     B: Body,
 {
+    pub(crate) fn new(inner: BodyInner<B>) -> Self {
+        Self { inner }
+    }
+
     /// Get a reference to the inner body
     pub fn get_ref(&self) -> &B {
-        match &self.0 {
+        match &self.inner {
             #[cfg(feature = "compression-gzip")]
-            BodyInner::Gzip(inner) => inner.read.get_ref().get_ref().get_ref().get_ref(),
+            BodyInner::Gzip { inner } => inner.read.get_ref().get_ref().get_ref().get_ref(),
             #[cfg(feature = "compression-deflate")]
-            BodyInner::Deflate(inner) => inner.read.get_ref().get_ref().get_ref().get_ref(),
+            BodyInner::Deflate { inner } => inner.read.get_ref().get_ref().get_ref().get_ref(),
             #[cfg(feature = "compression-br")]
-            BodyInner::Brotli(inner) => inner.read.get_ref().get_ref().get_ref().get_ref(),
-            BodyInner::Identity(inner) => inner,
+            BodyInner::Brotli { inner } => inner.read.get_ref().get_ref().get_ref().get_ref(),
+            BodyInner::Identity { inner } => inner,
+
+            #[cfg(not(feature = "compression-gzip"))]
+            BodyInner::Gzip { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-deflate"))]
+            BodyInner::Deflate { inner } => match inner.0 {},
+            #[cfg(not(feature = "compression-br"))]
+            BodyInner::Brotli { inner } => match inner.0 {},
```

Comment:
this certainly _feels_ weird to a reader who doesn't know what's going on here...it might be worth adding a comment explaining why this code is necessary, with a reference to https://github.com/rust-lang/rust/issues/51085

but, not a blocker.